### PR TITLE
[RISCV][MRI] Account for fixed registers when determining callee saved regs

### PR DESCRIPTION
**Pull request by michaelmaitland**
This fixes https://discourse.llvm.org/t/fixed-register-being-spill-and-restored-in-clang/83058.

We need to do it in `MachineRegisterInfo::getCalleeSavedRegs` instead of `RISCVRegisterInfo::getCalleeSavedRegs` since the MF argument of `TargetRegisterInfo:::getCalleeSavedRegs` is `const`, so we can't call `MF->getRegInfo().disableCalleeSavedRegister` there.

So to put it in `MachineRegisterInfo::getCalleeSavedRegs`, we move `isRegisterReservedByUser` into `TargetSubtargetInfo`.

### CODE REVIEW

**Comment by topperc** [comment]
Does this change the stack layout of callee saved registers when one is reserved? Does it match gcc stack layout?

**Comment by michaelmaitland** [comment]
> Does this change the stack layout of callee saved registers when one is reserved? 

This patch causes us not to spill the reserved register, when we were spilling it before. This by definition changes the stack layout.

> Does it match gcc stack layout?

I tried with this program with my patch applied:

```
register long a asm("x24");
int bar();
int baz(int a, int b, int c, int d);
int foo(int b, int c, int d) {
  a = 123;
  bar();
  baz(a, b, c, d);
  return 0;
}
```

GCC:
```
sd      ra,24(sp)
sd      s0,16(sp)
sd      s1,8(sp)
sd      s2,0(sp)
```
LLVM + this patch:
```
sd ra, 24(sp) # 8-byte Folded Spill
sd s0, 16(sp) # 8-byte Folded Spill
sd s1, 8(sp) # 8-byte Folded Spill
sd s2, 0(sp) # 8-byte Folded Spill
```

Looks the same on this example.

**Comment by topperc** [comment]
> > Does this change the stack layout of callee saved registers when one is reserved?
> 
> This patch causes us not to spill the reserved register, when we were spilling it before. This by definition changes the stack layout.
> 
> > Does it match gcc stack layout?
> 
> I tried with this program with my patch applied:
> 
> ```
> register long a asm("x24");
> int bar();
> int baz(int a, int b, int c, int d);
> int foo(int b, int c, int d) {
>   a = 123;
>   bar();
>   baz(a, b, c, d);
>   return 0;
> }
> ```
> 
> GCC:
> 
> ```
> sd      ra,24(sp)
> sd      s0,16(sp)
> sd      s1,8(sp)
> sd      s2,0(sp)
> ```
> 
> LLVM + this patch:
> 
> ```
> sd ra, 24(sp) # 8-byte Folded Spill
> sd s0, 16(sp) # 8-byte Folded Spill
> sd s1, 8(sp) # 8-byte Folded Spill
> sd s2, 0(sp) # 8-byte Folded Spill
> ```
> 
> Looks the same on this example.

A better test would spill s7 and s9, but not s8. I want to know if s8 still gets allocated a slot in gcc.

**Comment by michaelmaitland** [comment]
It looks like we are still allocating a slack slot.
```
register long a asm("x8");

int foo(int b, int c) {
  asm volatile ("" ::: "x9");
  a = 321;
  return 0;
}
```
GCC:
```
        addi    sp,sp,-16
        sd      s1,8(sp)
        li      s0,321
        ld      s1,8(sp)
        li      a0,0
        addi    sp,sp,16
        jr      ra
```
LLVM + this patch:
```
          addi    sp, sp, -16
        sd      s0, 8(sp)                       # 8-byte Folded Spill
        sd      s1, 0(sp)                       # 8-byte Folded Spill
        li      s0, 321
        li      a0, 0
        ld      s0, 8(sp)                       # 8-byte Folded Reload
        ld      s1, 0(sp)                       # 8-byte Folded Reload
        addi    sp, sp, 16
        ret
```

**Comment by topperc** [comment]
> LLVM + this patch:
> 
> ```
>         addi    sp, sp, -16
>         sd      s0, 8(sp)                       # 8-byte Folded Spill
>         sd      s1, 0(sp)                       # 8-byte Folded Spill
>         li      s0, 321
>         li      a0, 0
>         ld      s0, 8(sp)                       # 8-byte Folded Reload
>         ld      s1, 0(sp)                       # 8-byte Folded Reload
>         addi    sp, sp, 16
>         ret
> ```

Shouldn't this patch prevent the save/restore of s0?

**Comment by jrtc27** [comment]
Reserving x8 doesn't seem like a sensible thing to do, that's one of the few registers that really should be under the compiler's control, no?

**Comment by michaelmaitland** [comment]
> > LLVM + this patch:
> > ```
> >         addi    sp, sp, -16
> >         sd      s0, 8(sp)                       # 8-byte Folded Spill
> >         sd      s1, 0(sp)                       # 8-byte Folded Spill
> >         li      s0, 321
> >         li      a0, 0
> >         ld      s0, 8(sp)                       # 8-byte Folded Reload
> >         ld      s1, 0(sp)                       # 8-byte Folded Reload
> >         addi    sp, sp, 16
> >         ret
> > ```
> 
> Shouldn't this patch prevent the save/restore of s0?

Sorry, I collected this experiment without this patch (I had built another branch and forgot). I have updated the test cases now with the correct code. And I addressed @jrtc27 concern of using x8.

**Comment by michaelmaitland** [comment]
* `isRegisterReservedByUser` has the same behavior on both RISC-V and M68k. I think it makes sense to use a common interface.
* On LLVM, M68k target does not allow named registers in ASM. I've added a test to show this.
* [godbolt](https://godbolt.org/z/4n6sGTvoh) shows us that GCC does not save/restore for fixed registers. I think it makes sense for M68k target to eventually get the behavior in this patch.
* It is out of the scope of this PR to add named register support to M68k target, and they will have to modify the test added in this patch when they go to add this support.

File: llvm/test/CodeGen/M68k/fixed-csr.ll [hunk 1]
```diff
@@ -0,0 +1,12 @@
+; RUN: not --crash llc -mtriple=m68k -mattr=+reserve-a0 < %s 2>&1 | FileCheck %s
```
[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/test/CodeGen/M68k/fixed-csr.ll:1]: Ok, i've updated it to use callee saved register d2 and verified that gcc behavior is still as expected: https://godbolt.org/z/xnGPWE76M

**Comment by michaelmaitland** [comment]
The behavior differs from GCC for AArch64, ARM, Hexagon, SPARC, M68k and RISC-V. 

In LLVM, different targets have different interfaces to check whether the register was `-ffixed-XXX`. I am proposing that each target migrate to this common interface in follow up patches.

This patch adds the interface, and migrates RISC-V and M68k to that interface, since they are already using the interface.

**Comment by topperc** [comment]
> The behavior differs from GCC for AArch64, ARM, Hexagon, SPARC, M68k and RISC-V.
> 
> In LLVM, different targets have different interfaces to check whether the register was `-ffixed-XXX`. I am proposing that each target migrate to this common interface in follow up patches.
> 
> This patch adds the interface, and migrates RISC-V and M68k to that interface, since they are already using the interface.

How will this interact with AArch64 or other targets calling setCalleeSavedRegisters which will cause `IsUpdatedCSRsInitialized` to get set preventing reaching the new code you added in `getCalleeSavedRegs`

**Comment by michaelmaitland** [comment]
> > The behavior differs from GCC for AArch64, ARM, Hexagon, SPARC, M68k and RISC-V.
> > In LLVM, different targets have different interfaces to check whether the register was `-ffixed-XXX`. I am proposing that each target migrate to this common interface in follow up patches.
> > This patch adds the interface, and migrates RISC-V and M68k to that interface, since they are already using the interface.
> 
> How will this interact with AArch64 or other targets calling setCalleeSavedRegisters which will cause `IsUpdatedCSRsInitialized` to get set preventing reaching the new code you added in `getCalleeSavedRegs`

If a target calls `setCalleeSavedRegs`, then its saying "give me full control of what registers are callee saved regs".  If a target calls `setCalleeSavedRegs`, then it is on the target to determine which registers are fixed, in the same way its the target to figure out what registers are callee saved regs.

If a target calls `getCalleeSavedRegs`, it is prevented from reaching the existing `getTargetRegisterInfo()->getCalleeSavedRegs()` code. How is this any different?

**Comment by topperc** [comment]
> > > The behavior differs from GCC for AArch64, ARM, Hexagon, SPARC, M68k and RISC-V.
> > > In LLVM, different targets have different interfaces to check whether the register was `-ffixed-XXX`. I am proposing that each target migrate to this common interface in follow up patches.
> > > This patch adds the interface, and migrates RISC-V and M68k to that interface, since they are already using the interface.
> > 
> > 
> > How will this interact with AArch64 or other targets calling setCalleeSavedRegisters which will cause `IsUpdatedCSRsInitialized` to get set preventing reaching the new code you added in `getCalleeSavedRegs`
> 
> If a target calls `setCalleeSavedRegs`, then its saying "give me full control of what registers are callee saved regs". If a target calls `setCalleeSavedRegs`, then it is on the target to determine which registers are fixed, in the same way its the target to figure out what registers are callee saved regs.
> 
> If a target calls `getCalleeSavedRegs` and IsUpdatedCSRsInitialized is true, it is prevented from reaching the existing `getTargetRegisterInfo()->getCalleeSavedRegs()` code. How is this any different?

The existing AArch64 code that uses `setCalleeSavedRegs` calls `AArch64RegisterInfo::getCalleeSavedRegs` and then adds some stuff to it. I guess the code you added to getCalleeSavedRegs will need to be copied into `AArch64RegisterInfo::UpdateCustomCalleeSavedRegs` to remove the fixed registers?

I'm just asking questions to make sure you've designed something that covers the use cases of other targets. If you think having AArch64 fix implement their fixed register support in `AArch64RegisterInfo::UpdateCustomCalleeSavedRegs` is the best solution, I'm ok with that.

**Comment by michaelmaitland** [comment]
> The existing AArch64 code that uses setCalleeSavedRegs calls AArch64RegisterInfo::getCalleeSavedRegs and then adds some stuff to it. I guess the code you added to getCalleeSavedRegs will need to be copied into AArch64RegisterInfo::UpdateCustomCalleeSavedRegs to remove the fixed registers?

I'm interpreting the existing code in AArch64 as copying the call of getTargetRegisterInfo->getCalleeSavedRegs out of `MachineRegisterInfo::getCalleeSavedRegs` and putting it into AArch64RegisterInfo::UpdateCustomCalleeSavedRegs. So yes, they would also need to copy the code to copy the code to remove the fixed registers. 

Or they could just call MachineRegisterInfo::getCalleeSavedRegs from the start, add some regs to the result, and call setCalleeSavedRegs. Or they could use getCalleeSavedRegs and add a function addCalleeSavedReg instead of using setCalleeSavedRegs.

**Comment by michaelmaitland** [comment]
ping

File: llvm/test/CodeGen/RISCV/fixed-csr.ll [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+; CHECK-NEXT:    li s8, 321
+; CHECK-NEXT:    li a0, 0
+; CHECK-NEXT:    ret
+  tail call void @llvm.write_register.i64(metadata !0, i64 321)
+  ret i32 0
+}
+
+declare void @llvm.write_register.i64(metadata, i64)
+
+define noundef signext i32 @bar() {
```
Comment (arichardson) [inline llvm/test/CodeGen/RISCV/fixed-csr.ll:16]: could add `nounwind` here to remove the .cfi directives from the test.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**